Behavior of non-versioned and versioned cluster tasks (#45917)
openshift/openshift-docs: modules/op-advantages-and-disadvantages-of-non-versioned-and-versioned-cluster-tasks.adoc

// This module is part of the following assembly:
//
// *cicd/pipelines/managing-nonversioned-and-versioned-cluster-tasks.adoc
:_content-type: CONCEPT
[id="advantages-and-disadvantages-of-non-versioned-and-versioned-cluster-tasks_{context}"]
= Advantages and disadvantages of non-versioned and versioned cluster tasks

> 

Reviewer: Hi @sounix000. Please add a sentence or short introductory paragraph to each of these files. 

From the OpenShift peer review checklist: "Each assembly and module must contain a sentence or short introductory paragraph after the H1 heading. This serves as the abstract for the file."

.Advantages and disadvantages of non-versioned and versioned cluster tasks
[options="header"]
|===

| Cluster task variant | Advantages | Disadvantages

> | Non-versioned cluster task 

Reviewer: Because you've given the name of the variant here, you can minimize the sentences on the same row that repeat it.

a| 

> * If you prefer {pipelines-shortname} with latest updates and bug fixes, use the non-versioned cluster task variant. 

Reviewer 2: `{pipelines-shortname}` what does this resolve to when the doc is rendered.

I'm asuming it is "pipelines" as in an instance of the CRD `Pipeline` (what customer use to setup their workflows). 🙂

Reviewer: ```suggestion
* If you prefer {pipelines-shortname} with the latest updates and bug fixes, use the non-versioned cluster task variant. 
```

> * If you prefer {pipelines-shortname} with latest updates and bug fixes, use the non-versioned cluster task variant. 

Reviewer: ```suggestion
* Gives {pipelines-shortname} with the latest updates and bug fixes. 
```
You don't need to repeat the category you've established in the first cell of the table.

> * The non-versioned variants are automatically upgraded during Operator upgrades, and  minimize redundant resources.  

Reviewer: ```suggestion
```

> * The non-versioned variants are automatically upgraded during Operator upgrades, and  minimize redundant resources.  

Reviewer: - Avoid passive voice.
```suggestion
* Upgrading the Operator upgrades non-versioned variants. This consumes fewer resources than having multiple versioned variants.
```

> a| {pipelines-shortname} using the non-versioned cluster tasks may break after an Operator upgrade, as the automatically upgraded cluster tasks may be backward incompatible. 

Reviewer: This statement belies the previous claim that "The metadata, specifications such as `params`, `workspaces`, and `steps`, and the behavior of both variants are same."

> a| {pipelines-shortname} using the non-versioned cluster tasks may break after an Operator upgrade, as the automatically upgraded cluster tasks may be backward incompatible. 

Reviewer: ```suggestion
a| {pipelines-shortname} using the non-versioned cluster tasks might break after an Operator upgrade if the automatically upgraded cluster tasks are backward-incompatible. 
```

| Versioned cluster task 
a| 
* If you prefer stable {pipelines-shortname} in production, use the versioned cluster task variant. 

> * The old version is retained on the cluster even after the new version of a cluster task is installed, which results in wastage of cluster resources.

Reviewer 2: this sounds like a disadvantage, doesn't it. 🤔

Reviewer: ```suggestion
* The old version is retained on the cluster even after the new version of a cluster task is installed, which wastes cluster resources.
```

a| 

> * If you continue using an old version of a cluster task, you may miss the latest features and critical security updates.

Reviewer: ```suggestion
* If you continue using an old version of a cluster task, you might miss the latest features and critical security updates.
```
Consider "may" a banned word. Consider using "might" or other alternatives instead. Search the IBM Style Guide for more information.

> * The old versions of cluster tasks that are not operational occupies resources on the cluster. 

Reviewer: ```suggestion
* The earlier versions of the cluster tasks that are not operational occupy resources on the cluster. 
```
For versions, use "earlier" and "later" instead of "older" and "newer."

> * You cannot remove the old versions of cluster tasks once the Operator is upgraded.  

Reviewer: ```suggestion
* You cannot remove the earlier versions of cluster tasks when the Operator is upgraded.  
```
Avoid using "once" to mean "when." IBM SG.

Reviewer: Consider making a list of troublesome words to avoid, such as "once", "may", and "as."

> * Once upgraded, the Operator cannot manage the old VCT. You can delete the old git VCT manually using the `oc delete clustertask` command, but you cannot restore them.  

Reviewer: ```suggestion
* When upgraded, the Operator cannot manage the old VCT. You can delete the old git VCT manually using the `oc delete clustertask` command, but you cannot restore them.  
```

.Advantages and disadvantages of non-versioned and versioned cluster tasks
[options="header"]
|===

| Cluster task | Advantages | Disadvantages

| Non-versioned cluster task (NVCT)
a| 
* If you prefer deploying pipelines with the latest updates and bug fixes, use the NVCT. 
* Upgrading the Operator upgrades the non-versioned cluster tasks, which consume fewer resources than multiple versioned cluster tasks.  
a| If you deploy pipelines that use NVCT, they might break after an Operator upgrade if the automatically upgraded cluster tasks are not backward-compatible. 

| Versioned cluster task (VCT)
a| 
* If you prefer stable pipelines in production, use the VCT. 
* The earlier version is retained on the cluster even after the later version of a cluster task is installed. You can continue using the earlier cluster tasks. 
a| 
* If you continue using an earlier version of a cluster task, you might miss the latest features and critical security updates.
* The earlier versions of cluster tasks that are not operational consume cluster resources.  

> * When upgraded , the Operator cannot manage the earlier VCT. You can delete the earlier VCT manually using the `oc delete clustertask` command, but you cannot restore it.  

Reviewer 3: extra space before comma (not sure if it would render correctly)

Reviewer 4: Eagle eyes, as always :) Thanks, Pavol. @sounix000 I will go ahead and merge this PR but please fix this as part of another PR.

Author: @ppitonak @Preeticp : Resolving this typo here: https://github.com/openshift/openshift-docs/pull/46294/files#r889220287. Thanks for merging it.
